src/Plutarch/Extra/Value.hs

```diff
@@ -170,3 +199,568 @@ phasOnlyOneTokenOfCurrencySymbol = phoistAcyclic $
+                                                # yHead
+                                    pure $
+                                        pif
+                                            (comp # needle # xEntry # yEntry)
+                                            (self # comp # xNext # pcon (PMap yNext))
+                                            (pcon PFalse)
+                                PNothing -> pure $ pcon PFalse
+                        )
+                        (pcon PTrue)
+                        (pto ys)
+            )
+
+{- |
+  Gets the first entry in the first item of the `PValue` mapping & returns the
+  rest of the PValue.
+
+  Fails if the PValue is empty. In cases where we know that a PValue contains
+  Ada, such as in the ScriptContext, then this will function will split the
+  Ada value - since the Ada entry comes first.
+-}
+psplitValue ::
+    forall (v :: AmountGuarantees) (s :: S).
+    Term
+        s
+        ( PValue 'Sorted v
+            :--> PPair
+                    (PAsData PInteger)
+                    (PValue 'Sorted v)
+        )
+psplitValue = phoistAcyclic $
+    plam $ \v ->
+        let vList ::
+                Term
+                    _
+                    ( PBuiltinList
+                        ( PBuiltinPair
+                            (PAsData PCurrencySymbol)
+                            (PAsData (PMap 'Sorted PTokenName PInteger))
+                        )
+                    )
+
+            vList = pto $ pto v
+         in pcon $
+                PPair
+                    ( psndBuiltin #$ phead #$ pto $
+                        pto $
+                            pfromData
+                                (psndBuiltin # (phead # vList))
+                    )
+                    (pcon $ PValue $ pcon $ PMap $ ptail # vList)
+
+{- |
+  Returns True if all entries in the first Value present in the second Value
+  are >=.
+-}
+pgeqValue ::
+    forall (v :: AmountGuarantees) (s :: S).
+    Term
+        s
+        ( PAsData (PValue 'Sorted v)
+            :--> PAsData (PValue 'Sorted v)
+            :--> PBool
+        )
+pgeqValue = phoistAcyclic $
+    plam $ \x y ->
+        x #== y
+            #|| pcmpMap # pgeqValueEntry # pto (pfromData x) # pto (pfromData y)
+
+{- |
+  Returns True if all entries in the first Map are >= the values in the second.
+-}
+pgeqValueEntry ::
+    forall (s :: S).
+    Term
+        s
+        ( PAsData PCurrencySymbol
+            :--> PAsData (PMap 'Sorted PTokenName PInteger)
+            :--> PAsData (PMap 'Sorted PTokenName PInteger)
+            :--> PBool
+        )
+pgeqValueEntry = phoistAcyclic $
+    plam $ \_sym x y ->
+        x #== y
+            #|| (pcmpMap # plam (const pgeqIntegerData) # pfromData x # pfromData y)
+  where
+    pgeqIntegerData = phoistAcyclic $ plam $ \a b -> pfromData b #<= pfromData a
+
+{- | `plookup` but also returning the tail of the list, and relying on
+ sorted order
+-}
```

Comment:
This needs a more meaningful name I feel. As well as a `since` and quote-links for the Haddock.